### PR TITLE
Optimize common functions with Neri-Schneider

### PR DESCRIPTION
**Pull request by DaniPopes**
Optimize `common.rs` by using Neri-Schneider algorithms, as presented in CppNow 2023: <https://www.youtube.com/watch?v=0s9F4QWAl-E>

There are likely more places where they can be applied, but I just wanted to open this PR for now :)

### CODE REVIEW

File: src/util/common.rs [hunk 1]
```diff
@@ -60,54 +74,36 @@ pub(crate) fn saturate_day_in_month(
...
 pub(crate) fn days_in_month(year: Year, month: Month) -> Day {
-    if month == 2 && is_leap_year(year) {
-        Day::N::<29>()
-    } else {
-        days_in_month_common_year(month)
-    }
+    Day::new_unchecked(days_in_month_unranged_unchecked(
+        year.get(),
+        month.get() as u8,
+    ))
```
Comment (BurntSushi) [inline src/util/common.rs:82]: You've gotta stay in ranged integer land here. As soon as you start creating new ranged integers like this, the constructor can't tell what the min/max values are allowed to be, and so it assumes the worst case.

There's a reason why there is duplicated code in this module. You can't use unranged code to implemented arithmetic for ranged integers, and you can't use ranged integers inside `const` functions.

Comment (DaniPopes) [reply src/util/common.rs]: Fixed in https://github.com/BurntSushi/jiff/pull/23/commits/bfefca72b020141c7478a2f877a0d9eff1857410.

FYI just adding `year.get()` in `days_in_month` panics with
```
thread 'civil::iso_week_date::tests::prop_next_day_is_greater' panicked at src/util/rangeint.rs:2173:1:
ri16 max 10000 is not in range -9999..=9999
```

File: src/util/common.rs [hunk 2]
```diff
@@ -60,54 +75,41 @@ pub(crate) fn saturate_day_in_month(
...
+    if month == 2 {
+        if is_leap_year(year) {
+            Day::N::<29>()
+        } else {
+            Day::N::<28>()
+        }
     } else {
-        days_in_month_common_year(month)
+        let month = month.get();
+        if (month ^ month >> 3) & 1 == 0 {
```
Comment (DaniPopes) [inline src/util/common.rs:88]: Do you want me to implement bit operations for range int?

File: src/util/common.rs [hunk 3]
```diff
@@ -1,53 +1,68 @@
...
 /// Returns true if and only if the given year is a leap year.
 ///
 /// A leap year is a year with 366 days. Typical years have 365 days.
 ///
 /// This doesn't used range integers and is thus `const`.
+#[inline]
 pub(crate) const fn is_leap_year_unranged(year: i16) -> bool {
-    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
+    let d = if year % 25 != 0 { 4 } else { 16 };
+    (year % d) == 0
```
Comment (scottlamb) [inline src/util/common.rs:57]: Why not check if `year` is an even multiple of 4/16 with masking operations instead? https://godbolt.org/z/P3xoqodfs The compiler does this sort of optimization when the dividend is constant, but in your version it's not.

Did you see a change in a benchmark here? I ran `cargo criterion` and didn't see any big changes between original / yours / my versions of `is_leap_year{,_unranged}`; not entirely sure there is a benchmark where this function is significant.

Comment (DaniPopes) [reply src/util/common.rs:57]: The compiler does not do any sort of optimization by default :) even at -Copt-level=1 they're identical

I don't expect them to show up in any generic criterion benchmark, you can check out the [CppNow 2023 talk](https://www.youtube.com/watch?v=0s9F4QWAl-E) for more details

Comment (scottlamb) [reply src/util/common.rs:57]: > The compiler does not do any sort of optimization by default :) even at -Copt-level=1 they're identical

I'm not sure what you're saying is identical, but it's not the two versions in my link. The assembly is substantially different.

> I don't expect them to show up in any generic criterion benchmark,

I'd suggest adding a specific one then to verify the changes are optimizations.

> you can check out the [CppNow 2023 talk](https://www.youtube.com/watch?v=0s9F4QWAl-E) at around 30 mins for more details on the optimizations done in this PR

You've pessimized the version from the talk. They did this:

```cpp
if (Y % 100 != 0)
    return Y % 4 == 0;

return Y % 16 == 0;
```

in which the division is by a constant and the compiler is easily able to strength-reduce the division. You factored the divisor into `d`, preventing it from doing so.

Comment (DaniPopes) [reply src/util/common.rs:57]: You're comparing the generated code in debug mode, which is just wrong; you have to pass `-Copt-level=1/2/3` to Godbolt to optimize it (3 is default for --release).

Here is mine compared to the talk's: https://godbolt.org/z/T4c5xn76o

I modified it from [datealgo-rs](https://github.com/nakedible/datealgo-rs/blob/a04855a9799b0ad4c06ba6d7baa8d97760b06a57/src/lib.rs#L819-L832), which is mentioned in the author's [eaf](https://github.com/cassioneri/eaf) repository

File: src/util/common.rs [hunk 4]
```diff
@@ -1,53 +1,68 @@
...
+to write `const`-compatible code, but I deemed it far too annoying to do
+so.
+
+# Algorithms
+
+Algorithms are taken from
+Neri C, Schneider L. "Euclidean affine functions and their application to calendar algorithms":
+- https://github.com/cassioneri/eaf/
+- https://www.youtube.com/watch?v=0s9F4QWAl-E
+*/
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply src/util/common.rs:36]: Must've forgot, will do, thank

File: src/util/common.rs (hunk 3 above)
Comment (scottlamb) [reply src/util/common.rs:58]: > You're comparing the generated code in debug mode, which is just wrong; you have to pass -Copt-level=1/2/3 to Godbolt to optimize it (3 is default for --release).

Oh! Yeah, I thought it defaulted to `--release`-like settings, but you're right, that's a huge difference.

**Pull request closed without merging**